### PR TITLE
FIX: PROPACK MKL compatibility

### PR DESCRIPTION
**Pull request by mckib2**
#### Reference issue

- gh-15108
- https://github.com/conda-forge/scipy-feedstock/pull/198

#### What does this implement/fix?

- defines macro instructing PROPACK to use a scipy g77 `[cz]dotc` wrappers
- incidental cleanup of `setup.py` file
 
#### Additional information

~~Use https://github.com/scipy/PROPACK/tree/scipy-czdotc-wrappers instead of scipy/PROPACK master~~ _already merged_

TODO:

- [x] should new wrappers be used always or only when g77 abi compat is asked for? _EDIT: yes_
- [x] complex128 still doesn't work correctly
- [x] update PROPACK submodule commit (when https://github.com/scipy/PROPACK/pull/3 is completed)
- [x] ilp64 support? _EDIT: nope_

### CODE REVIEW

**Review by rgommers — COMMENTED** [review]
Shouldn't this be using https://github.com/scipy/scipy/blob/master/scipy/_build_utils/src/wrap_dummy_g77_abi.f instead?

**Comment by rgommers** [comment]
I forgot the detailed history; looks like we got rid of `needs_g77_abi_wrapper` in other `setup.py` files. But using a slow routine that's duplicate of actual LAPACK functionality isn't ideal - maybe a temporary patch at best.

**Comment by mckib2** [comment]
> I forgot the detailed history; looks like we got rid of `needs_g77_abi_wrapper` in other `setup.py` files. But using a slow routine that's duplicate of actual LAPACK functionality isn't ideal - maybe a temporary patch at best.

This is actually how PROPACK is usually built: linked against the user's BLAS but uses custom LAPACK implementations.  I changed it to link against the user's LAPACK as well (noted in https://github.com/scipy/scipy/pull/14433#issuecomment-886010218).  I'm not sure going back to building PROPACK's vendored LAPACK implementations would help or not and it's certainly not what the user would expect.

I remember trying to use `needs_g77_abi_wrapper` in the original PR but didn't see anything change.  Wouldn't we have seen other ABI issues by now if this was an issue?  If not, I can certainly try

**Comment by tylerjereddy** [comment]
Maybe add a backport label when this is ready--I see PROPACK errors in CI test suite runs and some tricky debate above perhaps. This may require me to do an rc3 release when it is ready, fair enough.

**Comment by rgommers** [comment]
> I remember trying to use `needs_g77_abi_wrapper` in the original PR but didn't see anything change. Wouldn't we have seen other ABI issues by now if this was an issue? If not, I can certainly try

I remembered `*dotc` is special, it's one of only two functions special-cased in [wrap_dummy_g77_abi](https://github.com/scipy/scipy/blob/master/scipy/_build_utils/src/wrap_dummy_g77_abi.f). The commit message of commit acc253408531 has the explanation. That seemed to be macOS-specific though, so it may not be relevant.

**Comment by mckib2** [comment]
I tried replicating the steps taken in arpack's setup.py file and I can see no differences between compiling with `` sources and not.  The only thing I've found so far to prevent a crash on `cdotc` is using PROPACK's implementation.  I'm very happy to step through the code with you @rgommers if you have a few spare minutes later this week.

Hunting down the complex128 `L_not_one_step_loopgas_1` bug has brought me to this interesting backtrace:
```
Calling svds(a, k=1, solver='propack')
     75374:	/home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libiomp5.so: error: symbol lookup error: undefined symbol: ompt_start_tool (fatal)
     75374:	/home/mckib2/anaconda3/bin/python: error: symbol lookup error: undefined symbol: scalable_malloc (fatal)
     75374:	
     75374:	calling init: /home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libtbbmalloc.so.2
     75374:	
     75374:	find library=libmemkind.so [0]; searching
     75374:	 search path=/home/mckib2/anaconda3/bin/../lib		(RPATH from file /home/mckib2/anaconda3/bin/python)
     75374:	  trying file=/home/mckib2/anaconda3/bin/../lib/libmemkind.so
     75374:	 search cache=/etc/ld.so.cache
     75374:	 search path=/lib/x86_64-linux-gnu/tls/x86_64/x86_64:/lib/x86_64-linux-gnu/tls/x86_64:/lib/x86_64-linux-gnu/tls/x86_64:/lib/x86_64-linux-gnu/tls:/lib/x86_64-linux-gnu/x86_64/x86_64:/lib/x86_64-linux-gnu/x86_64:/lib/x86_64-linux-gnu/x86_64:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu/tls/x86_64/x86_64:/usr/lib/x86_64-linux-gnu/tls/x86_64:/usr/lib/x86_64-linux-gnu/tls/x86_64:/usr/lib/x86_64-linux-gnu/tls:/usr/lib/x86_64-linux-gnu/x86_64/x86_64:/usr/lib/x86_64-linux-gnu/x86_64:/usr/lib/x86_64-linux-gnu/x86_64:/usr/lib/x86_64-linux-gnu:/lib/tls/x86_64/x86_64:/lib/tls/x86_64:/lib/tls/x86_64:/lib/tls:/lib/x86_64/x86_64:/lib/x86_64:/lib/x86_64:/lib:/usr/lib/tls/x86_64/x86_64:/usr/lib/tls/x86_64:/usr/lib/tls/x86_64:/usr/lib/tls:/usr/lib/x86_64/x86_64:/usr/lib/x86_64:/usr/lib/x86_64:/usr/lib		(system search path)
     75374:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/tls/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64-linux-gnu/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/tls/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/tls/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/tls/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/tls/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64-linux-gnu/libmemkind.so
     75374:	  trying file=/lib/tls/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/lib/tls/x86_64/libmemkind.so
     75374:	  trying file=/lib/tls/x86_64/libmemkind.so
     75374:	  trying file=/lib/tls/libmemkind.so
     75374:	  trying file=/lib/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64/libmemkind.so
     75374:	  trying file=/lib/x86_64/libmemkind.so
     75374:	  trying file=/lib/libmemkind.so
     75374:	  trying file=/usr/lib/tls/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/tls/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/tls/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/tls/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/x86_64/libmemkind.so
     75374:	  trying file=/usr/lib/libmemkind.so
     75374:	
     75374:	/home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libiomp5.so: error: symbol lookup error: undefined symbol: omp_target_alloc_host (fatal)
     75374:	/home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libiomp5.so: error: symbol lookup error: undefined symbol: omp_target_alloc_shared (fatal)
     75374:	/home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libiomp5.so: error: symbol lookup error: undefined symbol: omp_target_alloc_device (fatal)
     75374:	/home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libiomp5.so: error: symbol lookup error: undefined symbol: omp_target_free (fatal)
[New Thread 0x7fffe81de780 (LWP 75378)]
[New Thread 0x7fffdfffe800 (LWP 75379)]
[New Thread 0x7fffe79dc880 (LWP 75380)]

Thread 4 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe79dc880 (LWP 75380)]
0x00007fffea67a238 in L_not_one_step_loopgas_1 () from /home/mckib2/anaconda3/lib/python3.9/site-packages/mkl/../../../libmkl_def.so.1
(gdb) 
```

Looks like the error might be with some missing libraries from the path -- still investigating.

**Comment by mckib2** [comment]
Force reinstalling `intel-openmp` as per this [stackoverflow answer](https://stackoverflow.com/questions/59764844/kernel-keeps-dying-and-restarting-when-running-some-lines-in-python-notebook) resolves the complex128 segfault (don't ask me why -- conda is a strange and mysterious beast to me, but it's probably something to do with multiple libraries providing the same symbol), now it's just the following exception:
```
capi_return is NULL
Call-back cb_zaprod_in___user__routines failed.
Traceback (most recent call last):
  File "/home/mckib2/Documents/scipy/ex.py", line 18, in <module>
    result = svds(a, k=1, solver='propack')
  File "/home/mckib2/Documents/scipy/scipy/sparse/linalg/_eigen/_svds.py", line 316, in svds
    res = _svdp(A, k=k, tol=tol**2, which=which, maxiter=None,
  File "/home/mckib2/Documents/scipy/scipy/sparse/linalg/_svdp.py", line 307, in _svdp
    u, sigma, bnd, v, info = lansvd(jobu, jobv, m, n, k, aprod, u, v, tol,
  File "/home/mckib2/Documents/scipy/scipy/sparse/linalg/_svdp.py", line 66, in __call__
    y[:] = self.A.matvec(x)
  File "/home/mckib2/Documents/scipy/scipy/sparse/linalg/_interface.py", line 230, in matvec
    raise ValueError('dimension mismatch')
ValueError: dimension mismatch
```

**Comment by rgommers** [comment]
> The only thing I've found so far to prevent a crash on `cdotc` is using PROPACK's implementation. I'm very happy to step through the code with you @rgommers if you have a few spare minutes later this week.

Thanks for trying. I'm traveling right now and quite short on time, so I'd prefer to not do this. I think this was enough effort spent on this; happy to go with whatever solution works (internal `cdotc` it seems).

**Comment by mckib2** [comment]
I have somehow over the last week lost my ability to reproduce this error.  Waiting for someone to confirm this is still an issue and/or how to reproduce an environment to get the issue

**Comment by mckib2** [comment]
Alright, with @WarrenWeckesser 's help I got the right conda environment with MKL installed and can reproduce again.

The changes in this PR (including changes to the PROPACK submodule) appear to fix this issue with MKL on Linux.  Can @rgommers or @WarrenWeckesser pull the changes (make sure to get new PROPACK changes!) and verify this resolves segfaults?  If so, should be good to merge and then we can see if it makes a difference for the other segfaults seen at https://github.com/conda-forge/scipy-feedstock/pull/198

**Comment by mckib2** [comment]
Last real CI failures looks like they're in [Main Windows Python38-32bit-fast](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=15625&view=logs&jobId=7d84eb14-5b07-5555-8e62-838a7038a2b0&j=7d84eb14-5b07-5555-8e62-838a7038a2b0&t=4554a219-d131-5036-f24a-790fdd69cbb3) and maybe [Main Linux_Python_38_32bit_full](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=15625&view=logs&jobId=7d84eb14-5b07-5555-8e62-838a7038a2b0&j=e3555eda-5afc-5d78-d2e9-cf9756f98375&t=220e105d-2d02-5073-9f3b-414490a3ddd5).  So we might have reduced the issues to problems on 32-bit platforms

**Comment by mckib2** [comment]
Well, good news is that the complex-valued issues seem to be resolved (as far as I can tell). 

Bad news is I'm not sure what's causing the crashes, but on a 32-bit VM I am able to see that some of the tests fail due to what appear to be precision/numerical issues.  E.g., recreating `sparse/linalg/_eigen/tests/test_svds.py::Test_SVDS_PROPACK::test_svds_input_validation_k_1[3]`:
```python
import numpy as np
from scipy.sparse.linalg import svds
rng = np.random.default_rng(0)
A = rng.random((4, 3))
svds(A, k=3, solver="propack")
```
gets the following exception:
```
numpy.linalg.LinAlgError:  An invariant subspace of dimension 2 was found.
```
It succeeds, however if `A` is cast to `np.single` precision.

We can replicate an example of a crash by giving the option `which="SM"` to the example above which yields the PROPACK Fortran error:
```
WARNING:  Invariant subspace found.  Dimension =             2
** On entry to DGEMM parameter number 10 had an illegal value
```
Similarly, run with single precision (`A.astype(np.single)`), there is no crash.

The pattern I'm starting to see (I'd have to investigate more to see if this is the only issue happening) is that double precision decompositions fail or crash on 32-bit machines while single precision decompositions succeed.

@rgommers Any thoughts about how to proceed here?

**Comment by mckib2** [comment]
Found an error in the `setup.py` where it was always building as if it were a 32-bit build.  This has been fixed in the latest push.

I also found that building without optimizations (`-O0`) for 32-bit solved the crashes encountered in the above examples, but not all tests are as lucky.  Still trying to figure out discriminators between examples/tests when they crash/raise/pass/etc.  Sparse matrices in particular are causing trouble -- still looking into it

**Comment by h-vetinari** [comment]
@mckib2 
Thanks a lot for working on this! Can you let me know if/when this PR is ready to be tested with the conda-forge infra?

**Comment by rgommers** [comment]
Nice, that sounds like progress.

> I also found that building without optimizations (`-O0`) for 32-bit solved the crashes encountered in the above examples

That seems okay to commit and leave in for a while. 32-bit performance is not so important, and the differences won't be huge probably.

**Comment by mckib2** [comment]
> That seems okay to commit and leave in for a while. 32-bit performance is not so important, and the differences won't be huge probably.

Just realized I've been testing on an i386 VM.  I noticed that Julia does not [provide a binary](https://github.com/JuliaBinaryWrappers/PROPACK_jll.jl/tree/master/src/wrappers) for this architecture.  I don't know if matters or not, but I can try finding an i686 VM to see if it works any better.  I am confused why the 32-bit tests are failing/crashing after these changes -- AFAIK they are passing in current master and the only things that changed are:
- removing `second.F` implementations (should be safe for modern Fortran?)
- adding `wrap_dummy_g77_abi.f` (don't believe should clobber any existing symbols)

I even tried building the `Lapack_Utils` files and (while a little better) tests still crash and some examples I tried lead to lots of NaNs.

@rgommers If this patch fixes complex crashes for 64-bit, how do we feel about disabling PROPACK for all 32-bit builds (at least until we can find a solution)?  We were already disabling complex-valued PROPACK functions due to some of these same strange issues.  Also, if scipy/PROPACK/pull/3 looks good, we can merge that and point this back to scipy's PROPACK repo and get 1 step closer to testing out changes with conda-forge

**Comment by mckib2** [comment]
> @rgommers If this patch fixes complex crashes for 64-bit, how do we feel about disabling PROPACK for all 32-bit builds (at least until we can find a solution)? We were already disabling complex-valued PROPACK functions due to some of these same strange issues. Also, if scipy/PROPACK/pull/3 looks good, we can merge that and point this back to scipy's PROPACK repo and get 1 step closer to testing out changes with conda-forge

I might have spoken too soon -- reverting the `second.F` deletions and removing the extra `wrap_dummy_g77_abi.f` allows the 32-bit to build and pass all tests...  Adjusted to apply changes to both appropriately (fingers crossed)

**Comment by mckib2** [comment]
> @mckib2 Thanks a lot for working on this! Can you let me know if/when this PR is ready to be tested with the conda-forge infra?

@h-vetinari If the CI passes this time around, I think we might be ready to try conda-forge again.  I'm going to spin up my MKL environment with these changes and make sure it still works to be safe

**Comment by mckib2** [comment]
I'm not too familiar with how ilp64 builds work -- any hints?  I obviously did something wrong in the `setup.py`.  Relevant error from ilp64 CI build:

<details>

```
  D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sblasext.o: In function `psaxpy_':
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/sblasext.F:184: undefined reference to `saxpy_'
  D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sblasext.o: In function `pscopy_':
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/sblasext.F:176: undefined reference to `scopy_'
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/sritzvec.F:170: undefined reference to `lsame_'
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/sritzvec.F:203: undefined reference to `second_'
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/sritzvec.F:142: undefined reference to `lsame_'
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/sritzvec.F:171: undefined reference to `lsame_'
  D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\ssafescal.o: In function `ssafescal_':
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/ssafescal.F:52: undefined reference to `slascl_'
  D:\a\1\s/scipy/sparse/linalg/_propack/PROPACK/single/ssafescal.F:46: undefined reference to `slamch_'
  D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o: In function `wzdotc_':
  D:/a/1/s/scipy/_build_utils/src/wrap_dummy_g77_abi.f:22: undefined reference to `zdotc_'
  D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o: In function `wzdotu_':
  D:/a/1/s/scipy/_build_utils/src/wrap_dummy_g77_abi.f:30: undefined reference to `zdotu_'
  D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o: In function `wzladiv_':
  D:/a/1/s/scipy/_build_utils/src/wrap_dummy_g77_abi.f:44: undefined reference to `zladiv_'
  D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o: In function `wcdotc_':
  D:/a/1/s/scipy/_build_utils/src/wrap_dummy_g77_abi.f:6: undefined reference to `cdotc_'
  D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o: In function `wcdotu_':
  D:/a/1/s/scipy/_build_utils/src/wrap_dummy_g77_abi.f:14: undefined reference to `cdotu_'
  D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o: In function `wcladiv_':
  D:/a/1/s/scipy/_build_utils/src/wrap_dummy_g77_abi.f:37: undefined reference to `cladiv_'
  collect2.exe: error: ld returned 1 exit status
  error: Command "C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin\gfortran.exe -Wall -g -Wall -g -shared D:\a\1\s\build\temp.win-amd64-3.9\Release\a\1\s\build\src.win-amd64-3.9\scipy\sparse\linalg\_propack\_spropack-f2pywrappers_blas64.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\printstat.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sblasext.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sbsvd.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sgemm_ovwr.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sgetu0.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\slanbpro.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\slansvd.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\slansvd_irl.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\smgs.risc.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sreorth.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\sritzvec.o D:\a\1\s\build\temp.win-amd64-3.9\scipy\sparse\linalg\_propack\PROPACK\single\ssafescal.o D:\a\1\s\build\temp.win-amd64-3.9\a\1\s\scipy\_build_utils\src\wrap_dummy_g77_abi.o -LC:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\lib\gcc\x86_64-w64-mingw32\8.1.0 -LC:\hostedtoolcache\windows\Python\3.9.9\x64\libs -LC:\hostedtoolcache\windows\Python\3.9.9\x64\PCbuild\amd64 -Lbuild\temp.win-amd64-3.9 -o build\temp.win-amd64-3.9\Release\.libs\lib_spropac.E5DAWI7ARN4G3VLJWFVWSZSMU5VALYVL.gfortran-win_amd64.dll build\temp.win-amd64-3.9\Release\.libs\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll -Wl,--allow-multiple-definition -Wl,--output-def,build\temp.win-amd64-3.9\Release\lib_spropac.E5DAWI7ARN4G3VLJWFVWSZSMU5VALYVL.gfortran-win_amd64.def -Wl,--export-all-symbols -Wl,--enable-auto-import -static -mlong-double-64" failed with exit status 1
```


</details>

File: scipy/sparse/linalg/_propack/setup.py
Comment (mckib2) [inline scipy/sparse/linalg/_propack/setup.py:28]: @pv What would it look like to add support for ilp64 here?  Looking through a lot of the other ilp64 PRs, it appears that function signatures have been changed to use typedef'd integer types.  Would a lot of PROPACK sources need to be changed for that to happen here?  (not that it wouldn't be worth it or doable, I'm just not familiar with this and would need help formulating a strategy)

**Comment by h-vetinari** [comment]
> If the CI passes this time around, I think we might be ready to try conda-forge again. I'm going to spin up my MKL environment with these changes and make sure it still works to be safe

Sounds like not everything passed, but I'm gonna give it a whirl anyway. 🙃 

> we can merge that and point this back to scipy's PROPACK repo and get 1 step closer to testing out changes with conda-forge

BTW, the scipy-feedstock can handle unmerged PRs in submodules, as long as there's a commit in some fork to aim for.

**Comment by rgommers** [comment]
Argh, we forgot to merge this because it wasn't marked for `1.8.0`. And this also needs to be updated to include https://github.com/scipy/PROPACK/pull/4 now I believe.

@mckib2 can you confirm that that is correct?

**Comment by mckib2** [comment]
> @mckib2 can you confirm that that is correct?

Yes, updating to the PROPACK submodule module should accomplish this.  scipy/PROPACK#4 doesn't seem to have fixed the problem, but it is a good fix nonetheless.  If CI is still green on this after that, then I think it's good to go in

**Comment by tylerjereddy** [comment]
@mckib2 the circleci failure is `fatal: couldn't find remote ref master`

I'm tempted to merge this regardless--this is because you need to rebase to use the `main` target there I'm pretty sure, since our default branch name has changed.

If you do feel confident rebasing that's good too--obviously, just be careful since it is pretty easy to mess up submodule updates when you do something big like a rebase. I think you know what I mean after working on submodules a bunch..

**Comment by tylerjereddy** [comment]
Either way, I need to try to squeeze out rc4 this weekend if possible. I believe we agree this is the final PR we need to merge or?

**Comment by mckib2** [comment]
@tylerjereddy

> If you do feel confident rebasing that's good too--obviously, just be careful since it is pretty easy to mess up submodule updates when you do something big like a rebase. I think you know what I mean after working on submodules a bunch..

I think the easy thing to do here is to just merge main back into this branch -- rebase will indeed be more painful.  I'm fairly confident that's an alright strategy here -- if this causes problems, let me know

**Review by tylerjereddy — APPROVED** [review]
Alright, let's see if this helps us get around the conda-forge issues blocking the release process. thanks

**Comment by tylerjereddy** [comment]
I squashed merged even though you wrote nice commit messages, mostly for my own sanity when backporting.

**Comment by rgommers** [comment]
Merging this change seems to cause a tricky to fix issue with `git submodule update`, it's not happy about the `master` to `main` rename in the PROPACK repo. This fixed it for me:
```
git submodule deinit -f .
git submodule update --init --recursive
```

**Comment by tylerjereddy** [comment]
yeah, I saw some complaints locally too.. hopefully things are "ok" as far as release process CI goes (clean slates may be "ok" I hope?)

**Comment by mckib2** [comment]
We are definitely seeing some sharp edges with submodules here, but it seems like we're managing by being careful.  I'm hopeful things are/will be better by using them, but if we're seeing too many issues/dangers, I'm happy to support re-evaluation

@tylerjereddy PROPACK submodule commit in scipy/main is correct, it matches the latest in PROPACK/main

**Comment by mdhaber** [comment]
> Merging this change seems to cause a tricky to fix issue...

I just experienced this, but I was lucky to have noticed this post. If everyone is likely to run into this, would it be worth an email to the developer mailing list?

**Comment by rgommers** [comment]
Yes, I'll do that now.

**Pull request merged**
